### PR TITLE
add support for async functions.

### PR DESCRIPTION
**Pull request by ulrichSchreiner**
A Function is a block of code which can return an error if it is not
successful. To call this function, one can use MustSucceed which tells
nsq to invoke the handler until the function itself returns no error.

The publisher and consumer are both implemented by the using process so
this function can be used to async call a function. If your process is
running inside a cluster it is possible to distribute the invocation to
other instances.

I think we will need such a functionality for `metal-api` to asynchronously delete machines (or other independent entities).

### CODE REVIEW

**Comment by ulrichSchreiner** [comment]
it's not ready yet.

File: bus/functools.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+func (f *Function) receive(par interface{}) error {
+	fp := par.(*funcParam)
+	return f.fn(fp.Parameter)
+}
+
+// MustSucceed invokes the function with no limit. So nsq will invoke the connected go function
+// until no error is returned.
+func (f *Function) MustSucceed(arg interface{}) error {
```
Comment (droid42) [inline bus/functools.go]: Isn't this dangerous to only offer "unlimited" retries?

File: bus/functools_test.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	if err != nil {
+		t.Fatalf("function must succeed, %v", err)
+	}
+
+	if res != value {
+		t.Errorf("result is %q, but should be %q", res, value)
+	}
+}
+
+func TestFunctionReplay(t *testing.T) {
```
Comment (droid42) [inline bus/functools_test.go]: I would suggest TestFunctionRetry

File: bus/functools_test.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+		wg.Done()
+		return nil
+	})
+	if err != nil {
+		t.Errorf("cannot create function, %v", err)
+	}
+
+	value := "Hello world"
+	err = f.MustSucceed(value)
+	wg.Wait()
```
Comment (droid42) [inline bus/functools_test.go]: Of course this sync on a waitgroup is only for this test and works well in this setting with only one function, but I think we should explicitly make clear that this is not a pattern that will work in a distributed setting.

File: go.mod [hunk 4]
```diff
@@ -26,6 +26,7 @@ require (
 	github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd // indirect
 	github.com/modern-go/reflect2 v1.0.1 // indirect
 	github.com/nsqio/go-nsq v1.0.8
+	github.com/nsqio/nsq v1.2.0
 	github.com/pkg/errors v0.9.1
 	github.com/pquerna/cachecontrol v0.0.0-20180517163645-1555304b9b35 // indirect
 	github.com/stretchr/testify v1.4.0
```
Comment (droid42) [inline go.mod]: Some updates are available

```
+----------------------------------------+------------------------------------+------------------------------------+--------+------------------+
|                 MODULE                 |              VERSION               |            NEW VERSION             | DIRECT | VALID TIMESTAMPS |
+----------------------------------------+------------------------------------+------------------------------------+--------+------------------+
| github.com/emicklei/go-restful         | v2.11.2+incompatible               | v2.12.0+incompatible               | true   | true             |
| github.com/emicklei/go-restful-openapi | v1.2.0                             | v1.3.0                             | true   | true             |
| github.com/metal-stack/v               | v1.0.1                             | v1.0.2                             | true   | true             |
| github.com/stretchr/testify            | v1.4.0                             | v1.5.1                             | true   | true             |
| go.uber.org/zap                        | v1.13.0                            | v1.14.1                            | true   | true             |
| golang.org/x/crypto                    | v0.0.0-20200214034016-1d94cc7ab1c6 | v0.0.0-20200317142112-1b76d66859c6 | true   | true             |
| gopkg.in/yaml.v3                       | v3.0.0-20200121175148-a6ecf24a6d71 | v3.0.0-20200313102051-9f266ea9e77c | true   | true             |
+----------------------------------------+------------------------------------+------------------------------------+--------+------------------+

```

**Pull request closed without merging**